### PR TITLE
[class] Ensure 'cast_class' is initialized for array type classes.

### PR DESCRIPTION
**Pull request by alexischr**
Fixes https://github.com/mono/mono/issues/9507

Background: Monoclass:cast_class is used internally to allow for some interfaces to be assignable to each other if they share a "storage class" instead of the class's element class. This is used to i.e. implement IList<ulong> being assignable from IList<long> (IList<ulong> has cast_class=long)

Consider a GenericClass<T> with a T[] field. On instantiation with e.g. GenericClass<ulong> , the ulong[] array class will be generated. As above, ulong[] has cast_class=long
When mono_class_init() is called on an array class, its element_class is also initialized.
If GenericClass<T>:Method() uses any type/subtype checking (e.g. if (field is double[]) ), those may attempt to use cast_class, which is not ensured as initialized.

### CODE REVIEW

**Comment by lambdageek** [comment]
If it works, I like this one better than #10626 - initializing `cast_class` right when the array type is initialized is less surprising than initializing a `cast_class` of a generic instantiation argument.

**Comment by kumpera** [comment]
Yeah, I agree that initing the element class makes more sense.

File: mono/metadata/class-init.c [hunk 1]
```diff
@@ -4159,10 +4159,16 @@ mono_class_init (MonoClass *klass)
...
 		MonoClass *element_class = klass->element_class;
+		MonoClass *cast_class = klass->cast_class;
+
 		if (!element_class->inited) 
 			mono_class_init (element_class);
 		if (mono_class_set_type_load_failure_causedby_class (klass, element_class, "Could not load array element class"))
 			goto leave;
+		if (!cast_class->inited)
+			mono_class_init (cast_class);
+		if (mono_class_set_type_load_failure_causedby_class (klass, element_class, "Could not load array cast class"))
```
Comment (lambdageek) [inline mono/metadata/class-init.c]:
```suggestion
		if (mono_class_set_type_load_failure_causedby_class (klass, cast_class, "Could not load array cast class"))
```

**Comment by alexischr** [comment]
Failures are unrelated flaky tests. Tests will now restart as I accept the commit for the typo. @lambdageek @kumpera I will merge if we're good

**Comment by lambdageek** [comment]
OSX failure in System.Runtime.InteropServices.HandleCollectorTests.System.Runtime.InteropServices.HandleCollectorTests.TestHandleCollector is #11132

**Pull request merged**